### PR TITLE
Spark: Add stored procedure API

### LINKED ISSUE #1593: Spark SQL Extensions: API for stored procedures
Introduce an API to plugin stored procedures.

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds the stored procedure API.

The actual implementation in Iceberg and tests will come in a separate PR.

Resolves #1593.

### CODE REVIEW

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+    // verify the number of params in the procedure match the number of params in the method
+    if (params.length != methodType.parameterCount) {
+      throw new AnalysisException("Method parameter count must match the number of procedure parameters")
+    }
+
+    // the MethodHandle API does not allow us to check the generic type
+    // so we only verify the return type is either void or iterable
+
+    if (outputType.nonEmpty && methodReturnType != classOf[java.lang.Iterable[_]]) {
```
Comment (aokolnychyi) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:82]: This is ugly. I'd appreciate any ideas.

Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:82]: In the current approach, `methodHandle` is expected to be either void or return `java.lang.Iterable` of Spark Rows.

File: spark3/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+package org.apache.spark.sql.connector.catalog;
+
+import java.lang.invoke.MethodHandle;
+import org.apache.spark.sql.types.StructType;
+
+public interface Procedure {
+  ProcedureParameter[] parameters();
+  StructType outputType();
+  MethodHandle methodHandle();
```
Comment (aokolnychyi) [inline spark3/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java:28]: Presto uses method handles for stored procedures too.

File: spark3-extensions/src/main/scala/org/apache/iceberg/spark/extensions/IcebergSparkSessionExtensions.scala [hunk 3]
```diff
@@ -20,13 +20,15 @@
...
 import org.apache.spark.sql.SparkSessionExtensions
+import org.apache.spark.sql.catalyst.analysis.ResolveProcedures
 import org.apache.spark.sql.catalyst.parser.extensions.IcebergSparkSqlExtensionsParser
 import org.apache.spark.sql.execution.datasources.v2.ExtendedDataSourceV2Strategy
 
 class IcebergSparkSessionExtensions extends (SparkSessionExtensions => Unit) {
 
   override def apply(extensions: SparkSessionExtensions): Unit = {
     extensions.injectParser { case (_, parser) => new IcebergSparkSqlExtensionsParser(parser) }
+    extensions.injectResolutionRule { _ => ResolveProcedures }
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/iceberg/spark/extensions/IcebergSparkSessionExtensions.scala:31]: The ignored argument is a `SparkSession`, which I think should be passed into `ResolveProcedures` so that it doesn't need to reference `SparkSession.active`.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+import org.apache.spark.sql.{AnalysisException, SparkSession}
+import org.apache.spark.sql.catalyst.expressions.{Expression, Literal}
+import org.apache.spark.sql.catalyst.plans.logical.{Call, CallArgument, CallStatement, LogicalPlan, NamedArgument, PositionalArgument}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.catalog.{CatalogManager, CatalogPlugin, LookupCatalog, Procedure, ProcedureCatalog, ProcedureParameter}
+import scala.collection.Seq
+
+object ResolveProcedures extends Rule[LogicalPlan] with LookupCatalog {
+
+  protected lazy val catalogManager: CatalogManager = SparkSession.active.sessionState.catalogManager
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:31]: The session should be passed in from rule injection.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala [hunk 5]
```diff
@@ -20,14 +20,21 @@
...
+import org.apache.spark.sql.catalyst.plans.logical.{Call, LogicalPlan}
 import org.apache.spark.sql.execution.SparkPlan
 
 object ExtendedDataSourceV2Strategy extends Strategy {
 
   override def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
-    case _: CallStatement =>
-      throw new AnalysisException("CALL statements are not currently supported")
+    case c @ Call(procedure, args) =>
+      CallExec(c.output, procedure.methodHandle, args.map(toScalaValue)) :: Nil
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala:32]: Should this be in `ExtendedDataSourceV2Strategy`? It doesn't seem related to DSv2.

File: spark3/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package org.apache.spark.sql.connector.catalog;
+
+import java.lang.invoke.MethodHandle;
+import org.apache.spark.sql.types.StructType;
+
+public interface Procedure {
+  ProcedureParameter[] parameters();
+  StructType outputType();
```
Comment (rdblue) [inline spark3/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java:27]: Does this need to be a struct?

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+    val argExprs = new Array[Expression](params.size)
+
+    nameToArgMap.foreach { case (name, arg) =>
+      val position = nameToPositionMap(name)
+      val param = params(position)
+      val paramType = param.dataType
+      val argType = arg.expr.dataType
+      if (paramType != argType) {
+        throw new AnalysisException(s"Wrong arg type for ${param.name}: expected $paramType but got $argType")
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:121]: This is going to force people to use value literals, like `12L`. Can we insert a cast and execute it if the type is compatible?

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+  private def buildNameToArgMap(
+      params: Seq[ProcedureParameter],
+      args: Seq[CallArgument],
+      nameToPositionMap: Map[String, Int]): Map[String, CallArgument] = {
+
+    val containsNamedArg = args.exists(_.isInstanceOf[NamedArgument])
+    val containsPositionalArg = args.exists(_.isInstanceOf[PositionalArgument])
+
+    if (containsNamedArg && containsPositionalArg) {
+      throw new AnalysisException("Named and positional arguments cannot be mixed")
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:146]: This is Presto's current behavior?

**Comment by rdblue** [comment]
Mostly looks good, but I would probably pass arguments using an interface that passes InternalRow rather than using MethodHandle.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala (hunk 7 above)
Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:121]: I think we can try to use `Cast$canUpCast` for this.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala (hunk 8 above)
Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:146]: Yes, Presto does not allow mixing named and positional args.

File: project/scalastyle_config.xml [hunk 9]
```diff
@@ -85,11 +85,6 @@
             <parameter name="maxTypes"><![CDATA[30]]></parameter>
         </parameters>
     </check>
-    <check level="error" class="org.scalastyle.scalariform.CyclomaticComplexityChecker" enabled="true">
```
Comment (aokolnychyi) [inline project/scalastyle_config.xml:88]: We've discussed how handy this rule is multiple times. I'd remove it if there are no objections.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala (hunk 7 above)
Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:85]: I added a cast if it is safe.

Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:85]: All expressions will undergo optimizations so we will get rid of casts before creating an exec node.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CallExec.scala [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.connector.catalog.Procedure
+
+case class CallExec(
+    output: Seq[Attribute],
+    procedure: Procedure,
+    input: InternalRow) extends V2CommandExec {
+
+  override protected def run(): Seq[InternalRow] = {
+    procedure.call(input)
```
Comment (aokolnychyi) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CallExec.scala:32]: @rdblue, I do not pass a type alongside the row since the rules rely on `parameters` reported by the procedure to align input expressions.

File: spark3/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java (hunk 6 above)
Comment (aokolnychyi) [reply spark3/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java:27]: Procedures are supposed to return Spark rows that will be reported as output. So it is the type of those rows.

File: spark3-extensions/src/main/scala/org/apache/iceberg/spark/extensions/IcebergSparkSessionExtensions.scala (hunk 3 above)
Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/iceberg/spark/extensions/IcebergSparkSessionExtensions.scala]: Done.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala (hunk 1 above)
Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala]: Got rid of this.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala (hunk 4 above)
Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala]: Done.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala (hunk 5 above)
Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala]: @rdblue, I think this is debatable. It does depend on the catalog API part of DS V2 so I've put it here. Any suggestions?

File: project/scalastyle_config.xml (hunk 9 above)
Comment (aokolnychyi) [reply project/scalastyle_config.xml:88]: I can move this into a separate PR and disable in Java too.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+      if (paramType != argType) {
+        argExprs(position) = Cast(arg.expr, paramType)
+      } else {
+        argExprs(position) = arg.expr
+      }
+    }
+
+    // assign nulls to optional params that were not set
+    params.foreach {
+      case p if !p.required && !nameToArgMap.contains(p.name) =>
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:97]: We know from validation above that if the map doesn't contain an argument, that it must be required. Not a big deal to have the additional check, though.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Call.scala [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
+import org.apache.spark.sql.connector.catalog.Procedure
+import scala.collection.Seq
+
+case class Call(procedure: Procedure, args: Seq[Expression]) extends Command {
+  override def output: Seq[Attribute] = procedure.outputType.toAttributes
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Call.scala:27]: I think this needs to be a lazy val, or else the attributes will have different IDs every time it is called.

File: spark3/src/main/java/org/apache/spark/sql/connector/catalog/ProcedureParameter.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.connector.catalog;
+
+import org.apache.spark.sql.types.DataType;
+
+public interface ProcedureParameter {
```
Comment (rdblue) [inline spark3/src/main/java/org/apache/spark/sql/connector/catalog/ProcedureParameter.java:24]: I think for all of the connector interfaces, we should add clear Javadoc.

**Review by rdblue — APPROVED** [review]
This looks good to me, except for two things:
1. Plan node `output` should always be a `val` so that the attributes don't change.
2. Java interfaces in the connector API should have Javadoc (but can be added later).

File: spark3/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.connector.catalog;
+
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.types.StructType;
+
+public interface Procedure {
```
Comment (rdblue) [inline spark3/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java:25]: One more thing: we should also add a `describe` method so that we can show these correctly in plans.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Call.scala [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
+import org.apache.spark.sql.catalyst.util.truncatedString
+import org.apache.spark.sql.connector.catalog.Procedure
+import scala.collection.Seq
+
+case class Call(procedure: Procedure, args: Seq[Expression]) extends Command {
+  override lazy val output: Seq[Attribute] = procedure.outputType.toAttributes
+
+  override def simpleString(maxFields: Int): String = {
```
Comment (aokolnychyi) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Call.scala:30]: @rdblue, I implemented `simpleString` in `Call` and `CallExec` that use `description` in `Procedure`.

File: spark3/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java (hunk 14 above)
Comment (aokolnychyi) [reply spark3/src/main/java/org/apache/spark/sql/connector/catalog/Procedure.java:25]: Done.

File: spark3/src/main/java/org/apache/spark/sql/connector/catalog/ProcedureParameter.java (hunk 13 above)
Comment (aokolnychyi) [reply spark3/src/main/java/org/apache/spark/sql/connector/catalog/ProcedureParameter.java:24]: I'll add Javadoc in a follow-up PR. I need to document other places like `IcebergSparkSessionExtensions`. I want to have at least one procedure to be merged using these APIs.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Call.scala (hunk 12 above)
Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Call.scala]: Fixed.

**Comment by rdblue** [comment]
Looks great. I'm committing this so we can start building stored procedures for actions!

Thanks for getting this done!
